File: crates/gdext/src/commands/undo.rs
use serde_json::{Value, json};

use godot::classes::EditorInterface;
use godot::obj::Singleton;

use super::{get_root, pipe};
use crate::dispatcher::MainThreadDispatcher;

pub const TOOL_NAMES: &[&str] = &["undo", "redo"];

pub struct UndoCommands;

impl UndoCommands {
    pub fn new() -> Self {
        Self
    }
}

impl Default for UndoCommands {
    fn default() -> Self {
        Self::new()
    }
}

impl super::CommandHandler for UndoCommands {
    fn can_handle(&self, tool: &str) -> bool {
        TOOL_NAMES.contains(&tool)
    }
    fn handle<'a>(
        &'a self,
        tool: &'a str,
        args: &'a Value,
        d: &'a MainThreadDispatcher,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Value, String>> + Send + 'a>>
    {
        Box::pin(self.handle_undo_tool(tool, args, d))
    }
    fn group_name(&self) -> &str {
        "undo"
    }
    fn tool_names(&self) -> &[&str] {
        TOOL_NAMES
    }
}

impl UndoCommands {
    pub async fn handle_undo_tool(
        &self,
        tool: &str,
        args: &Value,
        d: &MainThreadDispatcher,
    ) -> Result<Value, String> {
        let a = args.clone();
        match tool {
            "undo" => pipe(d.submit(move || cmd_undo(&a)).await),
            "redo" => pipe(d.submit(move || cmd_redo(&a)).await),
            _ => Err(format!("Unknown undo tool: {}", tool)),
        }
    }
}

fn cmd_undo(_args: &Value) -> Value {
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let ur = EditorInterface::singleton().get_editor_undo_redo();
    let Some(ur_manager) = ur else {
        return json!({"error": "EditorUndoRedoManager not available"});
    };
    let history_id = ur_manager.get_object_history_id(&root);
    let Some(undo_redo) = ur_manager.get_history_undo_redo(history_id) else {
        return json!({"error": "No undo history for current scene"});
    };
    let mut ur_obj = undo_redo;
    if !ur_obj.has_undo() {
        return json!({"success": false, "hint": "Nothing to undo"});
    }
    let name = ur_obj.get_current_action_name().to_string();
    let result = ur_obj.undo();
    json!({"success": result, "action": name})
}

fn cmd_redo(_args: &Value) -> Value {
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let ur = EditorInterface::singleton().get_editor_undo_redo();
    let Some(ur_manager) = ur else {
        return json!({"error": "EditorUndoRedoManager not available"});
    };
    let history_id = ur_manager.get_object_history_id(&root);
    let Some(undo_redo) = ur_manager.get_history_undo_redo(history_id) else {
        return json!({"error": "No redo history for current scene"});
    };
    let mut ur_obj = undo_redo;
    if !ur_obj.has_redo() {
        return json!({"success": false, "hint": "Nothing to redo"});
    }
    // Align with cmd_undo: capture action name before executing redo.
    let name = ur_obj.get_current_action_name().to_string();
    let result = ur_obj.redo();
    json!({"success": result, "action": name})
}

File: extensions/gdext/src/commands/undo.cpp
// commands/undo.cpp — undo/redo tools

#include "cmd_utils.hpp"
#include "handler_registry.hpp"
#include <godot_cpp/classes/editor_interface.hpp>
#include <godot_cpp/classes/editor_undo_redo_manager.hpp>
#include <godot_cpp/classes/undo_redo.hpp>
using namespace godot;

namespace godot_mcp {
namespace {

Dictionary cmd_undo(const Dictionary &) {
    Node *root = get_root(); if (!root) return make_error("no scene");
    EditorUndoRedoManager *ur = EditorInterface::get_singleton()->get_editor_undo_redo();
    if (!ur) return make_error("EditorUndoRedoManager not available");
    int64_t history_id = ur->get_object_history_id(root);
    UndoRedo *undo_redo = ur->get_history_undo_redo((uint32_t)history_id);
    if (!undo_redo) return make_error("No undo history for current scene");
    if (!undo_redo->has_undo()) { Dictionary r; r["success"] = false; r["hint"] = "Nothing to undo"; return r; }
    bool ok = undo_redo->undo();
    String name = undo_redo->get_current_action_name();
    Dictionary r; r["success"] = ok; r["action"] = name; return r;
}
Dictionary cmd_redo(const Dictionary &) {
    Node *root = get_root(); if (!root) return make_error("no scene");
    EditorUndoRedoManager *ur = EditorInterface::get_singleton()->get_editor_undo_redo();
    if (!ur) return make_error("EditorUndoRedoManager not available");
    int64_t history_id = ur->get_object_history_id(root);
    UndoRedo *undo_redo = ur->get_history_undo_redo((uint32_t)history_id);
    if (!undo_redo) return make_error("No redo history for current scene");
    if (!undo_redo->has_redo()) { Dictionary r; r["success"] = false; r["hint"] = "Nothing to redo"; return r; }
    bool ok = undo_redo->redo();
    String name = undo_redo->get_current_action_name();
    Dictionary r; r["success"] = ok; r["action"] = name; return r;
}
} // namespace

void register_undo(HandlerRegistry &reg) {
    reg.register_tool("undo", cmd_undo);
    reg.register_tool("redo", cmd_redo);
}
} // namespace godot_mcp